### PR TITLE
Add `raw.line`

### PR DESCRIPTION
**Pull request by Dherse**
Adds a new sub-element called `raw.line` that allows access to individual lines of stylized code.

The `raw.line` contains the following properties:

- the line number (int, starting at 1)
- the raw text (string)
- the highlighted text (content)
- whether it's the first line (bool)
- whether it's the last line (bool)

Line of raw code can be either individually `show`-ruled or accessed on a synthesized field called `lines` on `raw`.

It allows the following:
````ts
#show raw: it => stack(dir: ttb, ..it.lines)
#show raw.line: it => {
  box(
    width: 100%,
    height: 1.75em,
    inset: 0.25em,
    fill: if calc.rem(it.number, 2) == 0 { luma(90%) } else { white },
    align(horizon, stack(
      dir: ltr,
      box(width: 15pt)[#it.number],
      it.body,
    ))
  )
}

```typ
#show raw.line: block.with(
  fill: luma(60%)
);

Hello, world!

= A heading for good measure
```
````

image

A big thank you to @laurmaedje and @Edhebi for the inspiration and the help 🎉

### CODE REVIEW

**Comment by sitandr** [comment]
Maybe we could rename `raw.line.line` to `raw.line.number`, like `enum.item.number`? `line.line` seems extremely ambiguous to me and a bit inconsistent.

**Comment by Dherse** [comment]
> Maybe we could rename `raw.line.line` to `raw.line.number`, like `enum.item.number`? `line.line` seems extremely ambiguous to me and a bit inconsistent.

I didn't think of that, I will rename it ;)

**Comment by laurmaedje** [comment]
What about about number/total instead of number/first/last? seems more flexible.

File: crates/typst-library/src/text/raw.rs [hunk 1]
```diff
@@ -402,28 +449,146 @@ impl PlainText for RawElem {
...
-fn highlight_themed<F>(
-    node: &LinkedNode,
+/// A stylized line of raw text.
+///
+/// This is a helper element that is synthesized by [`raw`]($raw) elements.
+///
+/// It allows you to access various properties of the line, such as the line
+/// number, the raw non-highlighted text, the highlighted tex, and whether it
+/// is the first or last line of the raw block.
+#[elem(name = "line", title = "Raw Text / Code line", Show, PlainText)]
```
Comment (laurmaedje) [inline crates/typst-library/src/text/raw.rs:459]:
```suggestion
#[elem(name = "line", title = "Raw Text / Code Line", Show, PlainText)]
```

File: tests/typ/text/raw-line.typ [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+```
+
+#show raw.line: it => {
+    box(stack(
+        dir: ltr,
+        box(width: 15pt)[#it.number],
+        it.body,
+    ))
+    linebreak()
+}
```
Comment (laurmaedje) [inline tests/typ/text/raw-line.typ:19]: could you change to 2 spaces of indent?

File: crates/typst-library/src/text/raw.rs [hunk 3]
```diff
@@ -402,28 +449,146 @@ impl PlainText for RawElem {
     }
 }
 
-/// Highlight a syntax node in a theme by calling `f` with ranges and their
-/// styles.
-fn highlight_themed<F>(
-    node: &LinkedNode,
+/// A stylized line of raw text.
```
[the remark this answers is not in the document]
Comment (Dherse) [reply crates/typst-library/src/text/raw.rs:452]: highlighted maybe more apt

**Comment by Dherse** [comment]
> What about about number/total instead of number/first/last? seems more flexible.

The main reason why I didn't do it was that (initially) I had a hard time doing it for typst code because I didn't want to manually split/iterate over the code to count the number of lines, but I think there is a nicer way, I'll give it a try.

**Comment by laurmaedje** [comment]
What is the nicer way? I think iterating the code once to count the lines should be pretty cheap.

**Comment by Dherse** [comment]
> What is the nicer way? I think iterating the code once to count the lines should be pretty cheap.

The nicer way would be doing it as you're iterating over the lines, but it turns out to not really work since I would need a second pass to update all of them anyway. So I added a `count` value with the max number of lines.

**Comment by laurmaedje** [comment]
Thanks once more!

**Pull request merged**